app/queues/reconciler/reconciler-worker.server.ts

```diff
@@ -0,0 +1,132 @@
+import { Worker, Queue } from 'bullmq';
+import { redis } from '~/lib/redis.server';
+import { executeChangeSet } from '~/lib/dns.server';
+import logger from '~/lib/logger.server';
+import DnsDbCompareStructureGenerator from './DnsDbCompareStructureGenerator.server';
+import Route53CompareStructureGenerator from './Route53CompareStructureGenerator.server';
+import {
+  createRemovedChangeSetFromCompareStructures,
+  createUpsertedChangeSetFromCompareStructures,
+} from './createChangeSetFromCompareStructures.server';
+import { getIsReconciliationNeeded, setIsReconciliationNeeded } from '~/models/system-state.server';
+import type { Change } from '@aws-sdk/client-route-53';
+
+// S3 limit for a ChangeSet
+const CHANGE_SET_MAX_SIZE = 1000;
+const reconcilerQueueName = 'reconciler';
+
+// Queue initialization
+export const reconcilerQueue = new Queue(reconcilerQueueName, {
+  connection: redis,
+});
+
+const createChangeSet = async (): Promise<Change[]> => {
+  const [dbStructure, route53Structure] = await Promise.all([
+    new DnsDbCompareStructureGenerator().generate(),
+    new Route53CompareStructureGenerator().generate(),
+  ]);
+
+  const changeSet = [
+    ...createRemovedChangeSetFromCompareStructures({ dbStructure, route53Structure }),
+    ...createUpsertedChangeSetFromCompareStructures({ dbStructure, route53Structure }),
+  ];
+
+  return changeSet;
+};
+
+/**
+ * NORMAL MODE
+ *
+ * Execute the complete changeSet at once
+ */
+const pushChangesBulk = async (changeSet: Change[]): Promise<boolean> => {
+  const recordSetsToPush = Math.min(CHANGE_SET_MAX_SIZE, changeSet.length);
+
+  logger.debug(
+    `Reconciler NORMAL MODE - Reconciler intends to push the following ${recordSetsToPush} changes`,
+    {
+      changeSet,
+    }
+  );
+
+  await executeChangeSet(changeSet.slice(0, CHANGE_SET_MAX_SIZE));
+
+  // Return boolean => Is additional reconciliation needed
+  return changeSet.length > CHANGE_SET_MAX_SIZE;
+};
+
+/**
+ * LIMP MODE
+ *
+ * Try each change in the set one by one, isolate the offending one
+ */
+const pushChangesIndividually = async (changeSet: Change[]) => {
+  for (const change of changeSet) {
+    try {
+      logger.debug(`Reconciler LIMP MODE - Reconciler intends to push the following change`, {
+        change,
+      });
+
+      await executeChangeSet([change]);
+    } catch (error) {
+      logger.error(`Reconciler LIMP MODE - the following single change failed`, {
+        change,
+        error,
+      });
+    }
+  }
+};
+
+const reconcilerWorker = new Worker(
+  reconcilerQueueName,
+  async () => {
+    // When BullMQ is running on multiple nodes (Docker swarm), we need to make sure our concurrency is 1
+    const activeJobs = (await reconcilerQueue.getJobs('active'))
+      // BullMQ bug, sometimes I get an array element with `undefined`, that should not be possible
+      .filter((v) => !!v);
+
+    if (activeJobs.length > 1) {
```

Comment:
I missed the explanation on this (or it's been lost in GitHub with the updated commit).  Why `> 1`?

Reply:
Normally, when a BullMQ worker is added, it will behave as single threaded ... will only execute one job at a time.
But, if we add multiple instances, i.e. our docker swarm, we have one worker per instance, taking jobs from the queue.
---
When you add a repeat job to a queue, it is a special thing. What it causes the queue system to keep adding (delayed) jobs when the repeat pattern/integer dictates it.
---
In theory, it would be possible, that a job is running long (more than the 2 mins job repeat time), so the second swarm node would also pick up the next scheduled job, causing the system to run two reconcilers in parallel.
---
For this reason, I'm asking the BullMQ system ... tell me how many active jobs are there (this includes the current job too, that we are in right now). If the answer is > 1, it means that there was a pre-existing job already running, when we were started ==> we must exit to inhibit concurrency

Reply:
I guess, I should copy-paste the above into the code :)